crates/revm/src/state_change.rs

```diff
@@ -51,11 +55,131 @@ pub fn post_block_balance_increments(
     balance_increments
 }
 
-/// Applies the pre-block call to the EIP-4788 beacon block root contract, using the given block,
+/// todo: temporary move over of constants from revm until we've migrated to the latest version
+pub const HISTORY_SERVE_WINDOW: usize = 8192;
+
+/// todo: temporary move over of constants from revm until we've migrated to the latest version
+pub const HISTORY_STORAGE_ADDRESS: Address = address!("25a219378dad9b3503c8268c9ca836a52427a4fb");
+
+/// Applies the pre-block state change outlined in [EIP-2935] to store historical blockhashes in a
+/// system contract.
+///
+/// If Prague is not activated, or the block is the genesis block, then this is a no-op, and no
+/// state changes are made.
+///
+/// If the provided block is the fork activation block, this will generate multiple state changes,
+/// as it inserts multiple historical blocks, as outlined in the EIP.
+///
+/// If the provided block is after Prague has been activated, this will only insert a single block
+/// hash.
+///
+/// [EIP-2935]: https://eips.ethereum.org/EIPS/eip-2935
+#[inline]
+pub fn apply_blockhashes_update<DB: Database + DatabaseCommit>(
+    chain_spec: &ChainSpec,
+    block_timestamp: u64,
+    block_number: u64,
+    db: &mut DB,
+) -> Result<(), BlockExecutionError>
+where
+    DB::Error: std::fmt::Display,
+{
+    // If Prague is not activated or this is the genesis block, no hashes are added.
+    if !chain_spec.is_prague_active_at_timestamp(block_timestamp) || block_number == 0 {
+        return Ok(())
+    }
+    assert!(block_number > 0);
+
+    // Create an empty account using the `From<AccountInfo>` impl of `Account`. This marks the
+    // account internally as `Loaded`, which is required, since we want the EVM to retrieve storage
+    // values from the DB when `BLOCKHASH` is invoked.
+    let mut account = Account::from(AccountInfo::default());
+
+    // HACK(onbjerg): This is a temporary workaround to make sure the account does not get cleared
+    // by state clearing later. This balance will likely be present in the devnet 0 genesis file
+    // until the EIP itself is fixed.
+    account.info.balance = U256::from(1);
```

Comment:
Does this create a genesis hash mismatch with other clients? Or any state root conflicts bc we have diff balance vs others in the state?